lib/elements/text.js

```diff
@@ -154,27 +155,68 @@ class TextPrompt extends Prompt {
 
   render() {
     if (this.closed) return;
-    if (!this.firstRender) {
-      if (this.outputError)
-        this.out.write(cursor.down(lines(this.outputError) - 1) + clear(this.outputError));
-      this.out.write(clear(this.outputText));
+    if (this.firstRender) {
+      this.out.write('\n' + cursor.up(1));
+      this.prevRowCount = 0;
     }
     super.render();
-    this.outputError = '';
+    let outputPostMessage = ``;
+    this.postMessage = ``;
 
-    this.outputText = [
+    this.outputPrompt = [
       style.symbol(this.done, this.aborted),
       color.bold(this.msg),
       style.delimiter(this.done),
-      this.red ? color.red(this.rendered) : this.rendered
+      ``
     ].join(` `);
-
-    if (this.error) {
-      this.outputError += this.errorMsg.split(`\n`)
-          .reduce((a, l, i) => a + `\n${i ? ' ' : figures.pointerSmall} ${color.red().italic(l)}`, ``);
+    this.outputText = this.red ? color.red(this.rendered) : this.rendered;
+
+    const outputLength = (
+      /* symbol length */
+      2 +
+      /* prompt */
+      this.msg.length +
+      /* delimiter */
+      3 +
+      /* input */
+      this.value.length
+    )
+
+    const rowCount = lines(` `.repeat(outputLength)) - 1;
+    const curRowLength = outputLength - (this.out.columns * rowCount);
+    const cursorGrossPos = this.cursor - this.value.length + outputLength;
+    const cursorRowCnt = lines(` `.repeat(cursorGrossPos)) - 1;
+    const cursorHorzPos = cursorGrossPos - (this.out.columns * cursorRowCnt);
+    const cursorRowOffset = rowCount - cursorRowCnt;
+
+    if (this.error || this.hint) {
+      this.postMessage = this.error ? this.errorMsg : this.hint;
+      outputPostMessage = this.postMessage.split(`\n`)
+        .map((l, i) => `${i > 0 ? ' ' : figures.pointerSmall} ${this.error ? color.red().italic(l) : color.gray().italic(l)}` + erase.lineEnd)
+        .join(`\n`);
+      if (this.hint) this.error = false;
     }
 
-    this.out.write(erase.line + cursor.to(0) + this.outputText + cursor.save + this.outputError + cursor.restore);
+    const postMessageRowCount = lines(this.postMessage);
+
+    this.out.write(
+      (this.prevRowCount > 0 ? cursor.up(this.prevRowCount) : ``) +
+      cursor.to(0) +
+      this.outputPrompt +
+      (this.placeholder ? cursor.save : ``) +
+      this.outputText +
+      (!this.placeholder ? cursor.save : ``) +
+      (curRowLength !== this.out.columns ? erase.lineEnd : ``) +
+      (this.postMessage ?
+        `\n${outputPostMessage}` +
+        erase.down() +
+        cursor.up(postMessageRowCount + cursorRowOffset) :
+          ``
+      ) +
+      cursor.to(cursorHorzPos)
+      // cursor.restore
```

Comment:
`cursor.restore` causes problems when printing on, or issuing a line feed on, the last line of the terminal.  It will not restore the cursor to the correct line, and it will throw off line counts.